Earthfile

```diff
@@ -12,48 +12,265 @@
+        python3-flake8-blind-except \
+        python3-flake8-builtins \
+        python3-flake8-class-newline \
+        python3-flake8-comprehensions \
+        python3-flake8-deprecated \
+        python3-flake8-docstrings \
+        python3-flake8-import-order \
+        python3-flake8-quotes \
+        python3-pip \
+        python3-pytest \
+        python3-pytest-cov \
+        python3-pytest-repeat \
+        python3-pytest-rerunfailures \
+        python3-rosdep \
+        python3-rosinstall-generator \
+        python3-setuptools \
+        python3-vcstool \
+        wget \
+        vim
+
+  # Define the username and key variables
+  ENV USERNAME spaceros-user
+  ENV HOME_DIR=/home/${USERNAME}
+  ENV SPACEROS_DIR=/opt/spaceros
+
+  # Create the spaceros directory
+  RUN mkdir --mode=777 -p ${SPACEROS_DIR}
+
+  # Create a spaceros user
+  RUN useradd -m ${USERNAME} && \
+    echo "${USERNAME}:${USERNAME}" | chpasswd && \
+    usermod --shell /bin/bash ${USERNAME} && \
+    usermod -aG sudo ${USERNAME} && \
+    mkdir -p /etc/sudoers.d && \
+    echo "${USERNAME} ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers.d/${USERNAME} && \
+    chmod 0440 /etc/sudoers.d/${USERNAME}
+  USER ${USERNAME}
+  WORKDIR ${SPACEROS_DIR}
+
+  # Update the OpenGL version
+  RUN sudo add-apt-repository ppa:kisak/kisak-mesa
+  RUN --mount=type=cache,mode=0777,target=/var/cache/apt,sharing=locked,id=cache_apt_cache \
+      --mount=type=cache,mode=0777,target=/var/lib/apt,sharing=locked,id=lib_apt_cache \
+      sudo apt update && sudo apt upgrade -y
+
+  # Create install location and copy in relevant scripts
+  COPY --chown ${USERNAME}:${USERNAME} --dir docker/scripts/ ${SPACEROS_DIR}
 
 repos-file:
   FROM +setup
-
-  # Copy inputs and generate a repos file
-  WORKDIR /root
-  COPY --dir scripts/ ./
   COPY excluded-pkgs.txt ./
   COPY spaceros-pkgs.txt ./
   COPY spaceros.repos ./
+  # This is a fresh image, so we do not need to exclude installed packages.
   RUN --no-cache sh scripts/generate-repos.sh \
                  --outfile ros2.repos \
                  --packages spaceros-pkgs.txt \
                  --excluded-packages excluded-pkgs.txt \
-                 --rosdistro $ROSDISTRO
+                 --exclude-installed false \
+                 --rosdistro ${ROS_DISTRO}
   RUN --no-cache python3 scripts/merge-repos.py ros2.repos spaceros.repos -o output.repos
-
-  # Save the generated .repos file
   SAVE ARTIFACT output.repos AS LOCAL ros2.repos
+
+spaceros-artifacts:
+  # we must run it in a separate container, so that downstream tasks can be cached if vcs file does not change
+  FROM +setup
+  USER ${USERNAME}
+  WORKDIR ${SPACEROS_DIR}
+
+  COPY ros2.repos ./
+  COPY excluded-pkgs.txt ./
+
+  # we run vcstool inside this task, because some packages in `ros2.repos` are not pinned and otherwise
+  # earthly won't pull latest changes
+  RUN mkdir src
+  RUN vcs import --shallow --retry 3 --input ros2.repos src
+  RUN vcs export --exact src > exact.repos
```

Comment:
Just so I know - is the problem that I raised in here https://github.com/eholum/space-ros/pull/1#discussion_r1681445294 fixed?

Reply:
Ah, thanks @xfiderek.

I re-added the cache breaking echo from before. As an alternative, we could do something like https://github.com/eholum/space-ros/commit/7318ba4b798e91cdb445bb741d3befee0b46c9ac#diff-2705f6016a64189abd14041e2f3e8dbb829553295d9088e4eb42eccb5e61bf47R135 which will let us choose when to cache or not cache this layer.

That said, this isn't the bottleneck in terms of speed, but if rebuilding over and over locally it might be nice to go with that approach because we can skip duplicating caches. What do you think?

Reply:
i left a comment on original issue.

summary here: I favour using default caching behaviour of earthly/docker (as this is what most people expect) and documenting the aspects of cache growth/not picking up external (to dockerfile) changes by clearing the cache.

Reply:
+1. I think we can leave it as is with the `--no-cache` option. As Ezra said, in the future we shouldn't really be cloning during builds anyway, but we can leave this as is and acknowledge that we will have large caches.